### ISSUE TITLE
RFC: selection manipulation 

### ISSUE THREAD

**Issue by dsizzle** [HaikuArchives/ArtPaint#413]
Ok, I’ve been working on getting Scale to work with selections for a while now, and it has me leaning really heavily towards this conclusion:

1. Layer manipulations (Translate, Rotate, Scale) should do nothing / be disabled if there is no selection. If you want to manipulate the whole layer, do “Select All” first. 
2. Canvas manipulations (same as above) should *ignore* selections and  *always* operate on the image as a whole. 

Thoughts?

**Comment by humdingerb** [comment]
Generelly, if you feel it's otherwise too complicated/impossible to implement, do it as you see fit. That said, my opinion to these 2 points:

> 1. Layer manipulations (Translate, Rotate, Scale, Cut, Copy) should do nothing / be disabled if there is no selection. If you want to manipulate the whole layer, do “Select All” first.

I don't quite understand, why this makes it easier. Couldn't the app just 'internally do a "Select all" ' if there's no selection?

> 2. Canvas manipulations (same as above) should ignore selections and always operate on the image as a whole.

If it can't be helped... :)
In general, it would be super consistent if any action applies to the selection only, and the whole layer/canvas if there is no selection.

But I'm pretty sure you've given it a lot of thought, and assuming we go with your 2 suggestions, I propose to:
a) Do your 2), i.e. the Canvas manipulators always work on the whole canvas (= all layers), ignoring selections.
b) Move all Layer manipulators to the Edit menu, as it'll always only work on the current selection of the current layer.

**Comment by dsizzle** [comment]
> I don't quite understand, why this makes it easier. Couldn't the app just 'internally do a "Select all" ' if there's no selection?

It could do a select all then a select none afterwards, but I think the less behind-the-scenes stuff, the better, right?

> In general, it would be super consistent if any action applies to the selection only, and the whole layer/canvas if there is no selection.

In general, I agree, but I just really can’t think of a reasonable use case where cutting a hole through all layers for an operation makes sense. It also poses some difficult challenges- for example, imagine if you make a selection at the canvas level, through multiple transparent layers. Then as you move, scale, or rotate it, you’d expect all the different layers of transparency and blending etc to still apply and work across the layers, no?  That’s *possible* but ArtPaint certainly doesn’t handle that now.

For Scale in particular, the reason this issue is complicated is because scaling the canvas should (and does) change the size of the entire image. But should scaling just a piece of the canvas, if it extends outside the boundaries of the current canvas, also make the image larger? 

But at the Layer level, scaling the layer, or any part of it, should not affect the canvas size at all - in my current implementation it does because there’s no way for the manipulator to tell the difference between a single layer and the canvas. This proposal solves it because layer operations require a selection - and if we move the layer operations to the edit menu (there’s already such a ticket anyway) then there’s no implied connection to a layer anyway.

**Comment by dsizzle** [comment]
> Then as you move, scale, or rotate it, you’d expect all the different layers of transparency and blending etc to still apply and work across the layers, no? That’s possible but ArtPaint certainly doesn’t handle that now.

This might not actually be true. It might work after you click Ok but it certainly doesn’t work in “preview” mode.

**Comment by dsizzle** [comment]
I think I figured out a way around the scaling canvas/layer issue so I might yet convince myself this isn't necessary... 

...but I do wonder what's the point of Select All if Select None behaves the same way...

**Comment by humdingerb** [comment]
> ...but I do wonder what's the point of Select All if Select None behaves the same way...

It's a shortcut for "whole-layer" operations by not having to "Select all" first. Admittedly, a very small shortcut... :) And it increases consistency, because add-ons also work on the whole layer if there's no selection.
Granted, "Select all" wouldn't even be needed, but we also should accommodate users' expectations, who may come from other apps. It's logical to have people looking for "Select all" if there's a "Select none". And as long as both produce the same result, all is fine. :)

Coming back to the added-SHIFT-applies-action-to-all-layers from above... What are your thoughts there?
If you deem it possible and desireable, instead of introducing the subfoldering (similar to the Copy/Cut commands) we could do it Tracker-style. That is, holding SHIFT modifies the menu item labels. (The +SHIFT shortcuts always work, of course.)
Yes, it's not fantastically discoverable, but ArtPaint is sufficiently complex an application that users can be trusted to consult the docs to learn about all features.

**Comment by humdingerb** [comment]
Yeah, I meant just exactly like Tracker's e.g. "Window" menu that has "Close _ALT+W_" change to "Close all _SHIFT+ALT+W_" when you hold SHIFT before clicking the menu.

Although: jscipione did try to get that change be live, i.e. while the menu is open. Sadly, it seems it still has issues and therefore wasn't merged yet. See https://review.haiku-os.org/c/haiku/+/2293/18
